### PR TITLE
[MRG, ENH] Use spatial_colors by default in Evoked.plot()

### LINKED ISSUE #10612: Use spatial_colors by default in Evoked.plot()
When using Evoked.plot(), in fact I **always** want to use `spatial_colors=True` as this feature is so simple, yet extremely helpful.

Problem is, on the first try I almost always **forget** to pass the respective parameter, so then I need to amend my code and re-run it.

Also I believe that many users are probably not aware of the feature, as … well, as it's not the default.

Could we not make it such that if we have sensor locations, we use `spatial_colors=True` by default?

We'd still keep the parameter to allow for manual adjustment, of course.

WDYT?

cc @sappelhoff @cbrnr @alexrockhill @adam2392

### PR DESCRIPTION
**Pull request by HuseyinOrkun**
#### Reference issue
Fixes #10612 with respect to [comment](https://github.com/mne-tools/mne-python/issues/10612#issuecomment-1120288827)

#### What does this implement/fix?
Adds a default 'auto' option for spatial_colors argument in Evoked.plot().
Which makes spatial_colors True if channel locations exists and False otherwise.
Changes docstring in plot_evoked() to include string
Adds option chek to plot_evoked for spatial_colors argument

#### Additional information
Any additional information you think is important.

### CODE REVIEW

File: mne/evoked.py [hunk 1]
```diff
@@ -358,7 +358,7 @@ def ch_names(self):
     def plot(self, picks=None, exclude='bads', unit=True, show=True, ylim=None,
              xlim='tight', proj=False, hline=None, units=None, scalings=None,
              titles=None, axes=None, gfp=False, window_title=None,
-             spatial_colors=False, zorder='unsorted', selectable=True,
+             spatial_colors='auto', zorder='unsorted', selectable=True,
              noise_cov=None, time_unit='s', sphere=None, *, highlight=None,
              verbose=None):
         return plot_evoked(
```
Comment (agramfort) [inline mne/evoked.py:364]: I suspect you need to update some docstrings

thx @HuseyinOrkun

File: mne/viz/evoked.py [hunk 2]
```diff
@@ -277,6 +277,14 @@ def _plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
...
 
     _check_option('gfp', gfp, [True, False, 'only'])
+    _check_option('spatial_colors', spatial_colors, [True, False, 'auto'])
+
+    # Check channel location
+    if spatial_colors == 'auto':
+        if _check_ch_locs(info=info):
+            spatial_colors = True
+        else:
+            spatial_colors = False
```
Comment (drammock) [inline mne/viz/evoked.py:287]: simpler:

```suggestion
        spatial_colors = _check_ch_locs(info=info)
```

File: mne/viz/evoked.py [hunk 3]
```diff
@@ -736,11 +744,12 @@ def plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
            Plot GFP for EEG instead of RMS. Label RMS traces correctly as such.
     window_title : str | None
         The title to put at the top of the figure.
-    spatial_colors : bool
+    spatial_colors : str | bool
```
Comment (drammock) [inline mne/viz/evoked.py:747]:
```suggestion
    spatial_colors : bool | 'auto'
```

**Review by drammock — COMMENTED** [review]
I made a couple suggestions below.  I also want to ask: what should happen when user passes `picks`?  For example, we probably don't want spatial_colors when user passes *just one or a few channels* as a pick.  But we probably *do* want spatial_colors when user passes `picks='mag'` or `picks='eeg'` (i.e., all channels of one type).  So I think setting `spatial_colors` automatically maybe should take into account `picks` as well.  @agramfort (or other devs) what is your opinion here?

File: mne/viz/evoked.py (hunk 2 above)
Comment (HuseyinOrkun) [reply mne/viz/evoked.py]: Thx, changed the code according to your suggestions.

File: mne/evoked.py (hunk 1 above)
Comment (HuseyinOrkun) [reply mne/evoked.py:364]: thx @agramfort. I checked but couldn't find which docstrings I still need to update. Could you give me some pointers?

**Comment by drammock** [comment]
style error: https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=21083&view=logs&j=c1ec3add-0fcd-52ea-530a-328457bbfb2e&t=056d958c-96b0-5ff4-dcb2-798466eed616&l=15

File: mne/viz/evoked.py [hunk 4]
```diff
@@ -736,11 +741,12 @@ def plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
            Plot GFP for EEG instead of RMS. Label RMS traces correctly as such.
     window_title : str | None
         The title to put at the top of the figure.
-    spatial_colors : bool
+    spatial_colors : bool | 'auto'
         If True, the lines are color coded by mapping physical sensor
         coordinates into color values. Spatially similar channels will have
         similar colors. Bad channels will be dotted. If False, the good
-        channels are plotted black and bad channels red. Defaults to False.
+        channels are plotted black and bad channels red. Can be 'auto', uses True if
```
Comment (agramfort) [inline mne/viz/evoked.py:748]:
```suggestion
        channels are plotted black and bad channels red. If 'auto' uses True if
```

**Review by agramfort — COMMENTED** [review]
@HuseyinOrkun you'll also need to add a line in latest.inc file to document the improvement.

**Comment by drammock** [comment]
before this gets merge, I'd still like to get other devs' opinion on this question:

> what should happen when user passes `picks`? For example, we probably don't want spatial_colors when user passes _just one or a few channels_ as a pick. But we probably _do_ want spatial_colors when user passes `picks='mag'` or `picks='eeg'` (i.e., all channels of one type). So I think setting `spatial_colors` automatically maybe should take into account the value of `picks` as well.

**Comment by drammock** [comment]
also, there are a lot of CI errors, here are links to a couple:

- https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=21091&view=logs&j=305851a9-a7bb-55db-0042-7e2b6f48aa1c&t=4d37777d-f36a-53aa-9217-6386d15dddcd&l=3240
- https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=21091&view=logs&j=c04f1cb7-a28e-5e8a-4074-704b3a6c72b1&t=c5a3e5f7-2b1e-57e1-93fb-c99e1ea8bb1a&l=1133

FYI our tests treat any *warning* as an error, unless the warning is explicitly caught/expected in the test. From a quick glance at the error summary linked above, it appears that you're raising a `RuntimeWarning`? That probably shouldn't happen... I'll look at your code now.

File: doc/changes/latest.inc [hunk 5]
```diff
@@ -24,7 +24,7 @@ Current (1.2.dev0)
 Enhancements
 ~~~~~~~~~~~~
 - EEGLAB files (saved as MAT versions less than v7.3) can now be imported with :func:`mne.io.read_raw_eeglab` without the optional dependency ``pymatreader`` (:gh:`11006` by `Clemens Brunner`_)
-
+-  Adds a default 'auto' option for spatial_colors argument in Evoked.plot() (:gh:`11017` by :newcontrib:`Hüseyin Orkun Elmas`)
```
Comment (drammock) [inline doc/changes/latest.inc:27]:
```suggestion
-  :meth:`mne.Evoked.plot()` has a new default value ``'auto'`` for its ``spatial_colors`` parameter, which will use spatial colors automatically if channel locations are available (:gh:`11017` by :newcontrib:`Hüseyin Orkun Elmas`)
```

Because this changes the default value of a parameter, this is not an enhancement, it is an API change. Please move to the API section.

**Review by drammock — COMMENTED** [review]
the RuntimeWarnings seen in the CI errors indicate that several of our tests *do not specify* a value for `spatial_colors` (so they assume the old default of `False`).  Fixed by changing the test to explicitly say `spatial_colors=False` probably?

The other errors you'll need to dig a bit deeper, not obvious to me why your changes here would cause the shape mismatches or the "setting array element with sequence" errors.

**Comment by hoechenberger** [comment]
> before this gets merge, I'd still like to get other devs' opinion on this question:
> 
> 
> 
> > what should happen when user passes `picks`? For example, we probably don't want spatial_colors when user passes _just one or a few channels_ as a pick. But we probably _do_ want spatial_colors when user passes `picks='mag'` or `picks='eeg'` (i.e., all channels of one type). So I think setting `spatial_colors` automatically maybe should take into account the value of `picks` as well.
> 
> 

plot_joint() always uses spatial colors, hence unless we want to implement an inconsistency, we should either use spatial colors in evoked.plot() by default (without further preconditions), or we should reconsider plot_joint() as well

**Comment by drammock** [comment]
Ok, I'm fine changing both of we have to. So then what would you recommend @hoechenberger? Do you think we should take picks into account or not?

**Comment by hoechenberger** [comment]
> So then what would you recommend @hoechenberger? Do you think we should take picks into account or not?

I think taking picks into account would create some sort of convolutional interaction between the two parameters. I do see your point thought that if you only have very few or a single trace, defaulting to colors may not be great.

How about introducing an arbitrary threshold of, say, 3 or 5 (or even 1?); and spatial_colors defaulting to None, which will act as True for n_chs > threshold, and as False otherwise?

**Comment by drammock** [comment]
> How about introducing an arbitrary threshold of, say, 3 or 5 (or even 1?); and spatial_colors defaulting to None, which will act as True for n_chs > threshold, and as False otherwise?

After thinking more about this and reading your ideas, I'm leaning towards one of 2 possibilities:

1. Ignore picks when auto-seeing sister colors
2. If picks is just one single channel, `spatial_colors="auto"` becomes False.

Anything else seems either too arbitrary or too much "magic"

**Comment by hoechenberger** [comment]
> If picks is just one single channel, spatial_colors="auto" becomes False.

I like this!

File: mne/viz/evoked.py [hunk 6]
```diff
@@ -217,6 +217,21 @@ def _plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
...
+    # If picks is greater than 1, set spatial colors
+    # True if channel locations exists False otherwise.
+    if spatial_colors == 'auto':
+        if picks and len(picks) == 1:
+
+            # If picks is not None and just one single channel
+            # spatial_colors="auto" becomes False
+            spatial_colors = False
+        else:
+            spatial_colors = _check_ch_locs(info=evoked.info)
```
Comment (hoechenberger) [inline mne/viz/evoked.py:234]: The comment doesn't match the code, no? I don't see a check for `len(picks) > 1` or `<= 1`.

File: mne/viz/ica.py [hunk 7]
```diff
@@ -1001,12 +1001,12 @@ def _plot_ica_overlay_evoked(evoked, evoked_cln, title, show):
     fig.suptitle(title)
     axes = axes.flatten() if isinstance(axes, np.ndarray) else axes
 
-    evoked.plot(axes=axes, show=False, time_unit='s')
+    evoked.plot(axes=axes, show=False, time_unit='s', spatial_colors=False)
```
Comment (agramfort) [inline mne/viz/ica.py:1004]: why are these necessary?

File: mne/viz/tests/test_evoked.py [hunk 8]
```diff
@@ -133,7 +133,7 @@ def test_plot_evoked():
 
     # test `gfp='only'`: GFP (EEG) and RMS (MEG)
     fig, ax = plt.subplots(3)
-    evoked.plot(gfp='only', time_unit='s', axes=ax)
+    evoked.plot(gfp='only', time_unit='s', axes=ax, spatial_colors=False)
```
Comment (agramfort) [inline mne/viz/tests/test_evoked.py:136]: the fact that you need to add this False in many places is a code smell that suggests it can break users code. Can you explain why you needed these?

File: mne/viz/evoked.py (hunk 6 above)
Comment (agramfort) [reply mne/viz/evoked.py:234]: I would use a private helper function here something like

`spatial_colors = _check_spatial_colors(info, spatial_colors)`

that would return True or False and that can be reused in any place necessary

File: mne/viz/ica.py (hunk 7 above)
Comment (hoechenberger) [reply mne/viz/ica.py:1004]: I believe the reason is that in the overlay plots, the only colors we want are for differentiating between before and after data cleaning; spatial colors would make it more difficult to see what has changed

File: mne/viz/tests/test_evoked.py (hunk 8 above)
Comment (HuseyinOrkun) [reply mne/viz/tests/test_evoked.py:136]: In some places, such as this one, the addition was unnecessary, and the lack of it did not break the test; I thought I should return the tests to the original states (since the argument was 'False' by default). In some places evoked.plot() is followed by _get_amplitudes(fig), which throws a VisibleDeprecationWarning caused by trying to create np. array from lists with different lengths. In one other line, there is an assertion about the highlighted areas of the plot, which probably fails spatial_colors arguments is 'True' rather than 'False'. I think because the tests were written for the 'False' default argument. I am not sure if this would break the user code.

File: mne/viz/ica.py (hunk 7 above)
Comment (HuseyinOrkun) [reply mne/viz/ica.py:1004]: Also, since the default was 'False' before, I thought not adding 'spatial_colors'=False would change the function's (_plot_ica_overlay_evoked_plot_ica_overlay_evoked) behavior.

File: mne/viz/tests/test_evoked.py (hunk 8 above)
Comment (agramfort) [reply mne/viz/tests/test_evoked.py:136]: ok it makes more sense to me now 🙏 

@HuseyinOrkun can I ask you to rebase this branch on current main branch and fix the conflicts to make the CIs green? 🙏

Comment (HuseyinOrkun) [reply mne/viz/tests/test_evoked.py:136]: I have a question, which is better, not changing the tests that do not break without the additional spatial_colors='False' argument or adding spatial_colors='False' to preserve the original arguments in those tests? Thank you very much for your feedback :)

File: doc/changes/latest.inc [hunk 9]
```diff
@@ -56,6 +56,8 @@ Bugs
 
 API changes
 ~~~~~~~~~~~
+- meth:`mne.Evoked.plot` has a new default value ``'auto'`` for its ``spatial_colors`` parameter, which will use spatial colors automatically if channel locations are available (:gh:`11017` by :newcontrib:`Hüseyin Orkun Elmas`)
```
Comment (agramfort) [inline doc/changes/latest.inc:59]: Line duplicated

**Review by agramfort — COMMENTED** [review]
Lets see if CIs are green.

**Comment by HuseyinOrkun** [comment]
Ready to be reviewed

File: doc/changes/latest.inc [hunk 10]
```diff
@@ -56,6 +56,7 @@ Bugs
...
```
Comment (hoechenberger) [inline doc/changes/latest.inc:59]:
```suggestion
- In meth:`mne.Evoked.plot`, the default value of the ``spatial_colors`` parameter has been changed to `'auto'`` for its ``, which will use spatial colors if channel locations are available (:gh:`11017` by :newcontrib:`Hüseyin Orkun Elmas`)
```

File: mne/viz/evoked.py [hunk 11]
```diff
@@ -193,6 +193,16 @@ def _plot_legend(pos, colors, axis, bads, outlines, loc, size=30):
     _draw_outlines(ax, outlines)
 
 
+def _check_spatial_colors(info, picks, spatial_colors):
+    """Set spatial colors True if channel locations exists, False otherwise."""
```
Comment (hoechenberger) [inline mne/viz/evoked.py:197]:
```suggestion
    """Use spatial colors if channel locations exist."""
```

File: mne/viz/evoked.py [hunk 12]
```diff
@@ -736,11 +748,13 @@ def plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
...
         The title to put at the top of the figure.
-    spatial_colors : bool
+    spatial_colors : bool | 'auto'
         If True, the lines are color coded by mapping physical sensor
         coordinates into color values. Spatially similar channels will have
         similar colors. Bad channels will be dotted. If False, the good
-        channels are plotted black and bad channels red. Defaults to False.
+        channels are plotted black and bad channels red. If 'auto' uses True if
+        channel locations are present, False if channel locations are not
+        present or there is only one channel. Defaults to 'auto'.
```
Comment (hoechenberger) [inline mne/viz/evoked.py:757]:
```suggestion
        channels are plotted black and bad channels red. If ``'auto'``, uses
        True if channel locations are present, and False if channel locations
        are missing or if the data contains only a single channel. Defaults to
        ``'auto'``.
```

**Review by hoechenberger — COMMENTED** [review]
Thanks, great work! I added a few suggestions for phrasing. Otherwise I think this looks good!

File: doc/changes/latest.inc [hunk 13]
```diff
@@ -56,6 +56,7 @@ Bugs
 
 API changes
 ~~~~~~~~~~~
+- In meth:`mne.Evoked.plot`, the default value of the ``spatial_colors`` parameter has been changed to `'auto'`` for its ``, which will use spatial colors if channel locations are available (:gh:`11017` by :newcontrib:`Hüseyin Orkun Elmas`)
```
Comment (hoechenberger) [inline doc/changes/latest.inc:59]:
```suggestion
- In meth:`mne.Evoked.plot`, the default value of the ``spatial_colors`` parameter has been changed to ``'auto'``, which will use spatial colors if channel locations are available (:gh:`11017` by :newcontrib:`Hüseyin Orkun Elmas`)
```

**Review by hoechenberger — APPROVED** [review]
